### PR TITLE
Refactor outreach opt-in to event driven state

### PR DESCRIPTION
**Pull request by nmajor25**
Shortcut Story ID: [sc-39934]

### CODE REVIEW

File: src/js/outreach/views/opt-in_views.js [hunk 1]
```diff
@@ -51,6 +51,9 @@ const OptInView = View.extend({
     <p class="opt-in__disclaimer">By clicking Submit you agree to receive SMS text message notifications. You may opt out at any time.</p>
     <button class="opt-in__submit button--green w-100 js-submit" disabled>Submit</button>
   `,
+  modelEvents: {
+    'change': 'setSubmitButtonState',
```
Comment (paulfalgout) [inline src/js/outreach/views/opt-in_views.js:57]: `setSubmitButtonState` should not use `this.ui.foo.val()` and should instead us `this.model.isValid()`. The data on the model is not necessarily the same data that's in the form.. `this.ui.phone.val` specifically.

**Review by paulfalgout — CHANGES_REQUESTED** [review]

File: src/js/outreach/views/opt-in_views.js (hunk 1 above)
Comment (nmajor25) [reply src/js/outreach/views/opt-in_views.js:57]: This has been updated in my latest push.

**Review by paulfalgout — APPROVED** [review]

File: src/js/outreach/views/opt-in_views.js [hunk 2]
```diff
@@ -69,25 +74,18 @@ const OptInView = View.extend({
...
-    this.setSubmitButtonState();
+    this.model.set({ last_name: trim(this.ui.lastName.val()) });
   },
   onChangeBirthDate() {
-    this.model.set({ birth_date: this.ui.birthDate.val() });
-    this.setSubmitButtonState();
+    this.model.set({ birth_date: trim(this.ui.birthDate.val()) });
   },
   onChangePhone() {
     const phone = parsePhoneNumber(this.ui.phone.val(), 'US');
```
Comment (nmajor25) [inline src/js/outreach/views/opt-in_views.js:86]: Don't think the phone number needs the `trim()`. Appears that `parsePhoneNumber()` handles all that for us.

`phone` will be `undefined` if the user enters only empty spaces. And strings like `"   12345   "` are converted to `"+112345"` automatically.

File: src/js/outreach/apps/opt-in_app.js [hunk 3]
```diff
@@ -18,6 +18,12 @@
     birth_date: '',
     phone: '',
   },
+  validate({ first_name, last_name, birth_date, phone }) {
+    if (!first_name) return 'invalid';
+    if (!last_name) return 'invalid';
+    if (!birth_date) return 'invalid';
+    if (!phone || phone === null) return 'invalid';
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply src/js/outreach/apps/opt-in_app.js]: @nmajor25 ah yeah. codeql caught this.  `null` is `!phone`

[the remark this answers is not in the document]
Comment (nmajor25) [reply src/js/outreach/apps/opt-in_app.js]: Hmmm so this should be `!phone.length || phone === null`?

We want to verify that `phone` isn't an empty string or equal to `null`.

[the remark this answers is not in the document]
Comment (nmajor25) [reply src/js/outreach/apps/opt-in_app.js]: Well `!phone` would just cover both...

[the remark this answers is not in the document]
Comment (paulfalgout) [reply src/js/outreach/apps/opt-in_app.js]: really this is the exact same check from the previous function `if (!firstName || !lastName || !birthDate || !phone) return 'invalid';`

[the remark this answers is not in the document]
Comment (paulfalgout) [reply src/js/outreach/apps/opt-in_app.js]: well I copied from the other thing which was camel case, but ya know

**Review by paulfalgout — APPROVED** [review]

File: src/js/outreach/apps/opt-in_app.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmajor25) [reply src/js/outreach/apps/opt-in_app.js]: Yup, exactly the same 👍 Alright that's been updated, hopefully all good now 🤞

**Pull request merged**